docs/howto/composite-primary-key.txt

```diff
@@ -0,0 +1,161 @@
+composite primary key.
+
+If you would like to migrate an existing table from a single primary key to a
+composite primary key, follow your database backend's instructions to do so.
+
+Once the composite primary key is in place, add the ``CompositePrimaryKey``
+field to your model. This allows Django to recognize and handle the composite
+primary key appropriately.
+
+While migration operations (e.g. ``AddField``, ``AlterField``) on primary key
+fields are not supported, ``makemigrations`` will still detect changes.
+
+In order to avoid errors, it's recommended to apply such migrations with
+``--fake``.
+
+Alternatively, :class:`.SeparateDatabaseAndState` may be used to execute the
+backend-specific migrations and Django-generated migrations in a single
+operation.
+
+.. _cpk-and-relations:
+
+Composite primary keys and relations
+====================================
+
+:ref:`Relationship fields <relationship-fields>`, including
+:ref:`generic relations <generic-relations>` do not support composite primary
+keys.
+
+For example, given the ``OrderLineItem`` model, the following is not
+supported::
+
+    class Foo(models.Model):
+        item = models.ForeignKey(OrderLineItem, on_delete=models.CASCADE)
+
+Because ``ForeignKey`` currently cannot reference models with composite primary
+keys.
+
+To work around this limitation, ``ForeignObject`` can be used as an
+alternative::
+
+    class Foo(models.Model):
+        item_order_id = models.IntegerField()
+        item_product_id = models.CharField(max_length=20)
+        item = models.ForeignObject(
+            OrderLineItem,
+            on_delete=models.CASCADE,
+            from_fields=("item_order_id", "item_product_id"),
+            to_fields=("order_id", "product_id"),
+        )
+
+``ForeignObject`` is much like ``ForeignKey``, except that it doesn't create
+any columns (e.g. ``item_id``), foreign key constraints or indexes in the
+database.
+
+.. warning::
+
+    ``ForeignObject`` is an internal API. This means it is not covered by our
+    :ref:`deprecation policy <internal-release-deprecation-policy>`.
+
+Composite primary keys and database functions
+=============================================
+
+Many database functions only accept a single expression.
+
+.. code-block:: sql
+
+    MAX("order_id")  -- OK
+    MAX("product_id", "order_id")  -- ERROR
+
+As a consequence, they cannot be used with composite primary key references as
+they are composed of multiple column expressions.
+
+.. code-block:: python
+
+    Max("order_id")  # OK
+    Max("pk")  # ERROR
+
+Composite primary keys in forms
+===============================
+
+:meth:`~django.db.models.Model.save()` doesn't support updating primary
+keys (composite or not).
+
+This means using :class:`~django.forms.ModelForm` with a composite
+primary key may lead to unexpected behavior.
+
+To avoid issues, make sure :attr:`~django.db.models.Field.editable`
+is set to ``False`` on all primary key fields.
```

Comment:
```suggestion
As a composite primary key is a virtual field, a field which doesn't represent
a single database column, this field is excluded from ModelForms::

    class OrderLineItemForm(forms.ModelForm):
        class Meta:
            model = OrderLineItem
            fields = "__all__"

.. code-block:: pycon

    >>> OrderLineItemForm()
    <OrderLineItemForm bound=False, valid=Unknown, fields=(product;order;quantity)>

Setting the primary composite field ``pk`` as a form field raises an unknown
field :exc:`.FieldError`.
```
I feel like forms work fine? I'm not sure the issue trying to be highlighted here. Can you confirm the behavior with tests?

Reply:
The way it works is `ModelForm` always excludes the primary key.

So yeah, they work. There's a test: `CompositePKTests.test_model_forms`.

The issue I tried to highlight is that the `ModelForm` doesn't exclude the fields of the composite primary key automatically. So those fields are editable on the form, which may lead to unexpected results.

Reply:
I see. I found a note on this within [Field.primary_key](https://docs.djangoproject.com/en/5.1/ref/models/fields/#django.db.models.Field.primary_key) (see the "The primary key field is read-only..")
So I have made this an admonition to reference that :+1: